File: app/src/main/java/me/devsaki/hentoid/activities/bundles/SearchActivityBundle.java
package me.devsaki.hentoid.activities.bundles;

import android.net.Uri;
import android.os.Bundle;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import me.devsaki.hentoid.database.domains.Attribute;
import me.devsaki.hentoid.enums.AttributeType;
import me.devsaki.hentoid.database.domains.AttributeMap;

/**
 * Helper class to transfer data from any Activity to {@link me.devsaki.hentoid.activities.SearchActivity}
 * through a Bundle.
 *
 * Use Builder class to set data; use Parser class to get data
 */
public class SearchActivityBundle {
    private static final String KEY_ATTRIBUTE_TYPES = "attributeTypes";
    private static final String KEY_MODE = "mode";
    private static final String KEY_URI = "uri";
    private static final String KEY_GROUP = "group";
    private static final String EXCLUDE_MODE = "exclude";

    private SearchActivityBundle() {
        throw new UnsupportedOperationException();
    }

    public static final class Builder {

        private final Bundle bundle = new Bundle();

        public void setAttributeTypes(AttributeType... attributeTypes) {
            ArrayList<Integer> attrTypes = new ArrayList<>();
            for (AttributeType type : attributeTypes) attrTypes.add(type.getCode());

            bundle.putIntegerArrayList(KEY_ATTRIBUTE_TYPES, attrTypes);
        }
        public void setExcludeMode(boolean excludeMode){
            bundle.putBoolean(EXCLUDE_MODE, excludeMode);
        }

        public void setMode(int mode) {
            bundle.putInt(KEY_MODE, mode);
        }

        public void setGroup(long group) {
            bundle.putLong(KEY_GROUP, group);
        }

        public Builder setUri(Uri uri) {
            bundle.putString(KEY_URI, uri.toString());
            return this;
        }

        public static Uri buildSearchUri(List<Attribute> attributes) {
            AttributeMap metadataMap = new AttributeMap();
            metadataMap.addAll(attributes);

            Uri.Builder searchUri = new Uri.Builder()
                    .scheme("search")
                    .authority("hentoid");


            for (Map.Entry<AttributeType, List<Attribute>> entry : metadataMap.entrySet()) {
                AttributeType attrType = entry.getKey();
                List<Attribute> attrs = entry.getValue();
                if (attrs != null)
                    for (Attribute attr : attrs)
                        searchUri.appendQueryParameter(attrType.name(), attr.getId() + ";" + attr.getName() + ";" + attr.isExcluded());
            }

            return searchUri.build();
        }

        public Bundle getBundle() {
            return bundle;
        }
    }

    public static final class Parser {

        private final Bundle bundle;

        public Parser(@Nonnull Bundle bundle) {
            this.bundle = bundle;
        }

        public List<AttributeType> getAttributeTypes() {
            List<AttributeType> result = new ArrayList<>();

            List<Integer> attrTypesList = bundle.getIntegerArrayList(KEY_ATTRIBUTE_TYPES);
            if (null != attrTypesList && !attrTypesList.isEmpty())
                for (Integer i : attrTypesList) result.add(AttributeType.searchByCode(i));

            return result;
        }
        public boolean getExcludeMode(){return bundle.getBoolean(EXCLUDE_MODE, false);}
        public int getMode() {
            return bundle.getInt(KEY_MODE, -1);
        }

        public long getGroupId() {
            return bundle.getLong(KEY_GROUP, -1);
        }

        @Nullable
        public Uri getUri() {
            Uri result = null;

            String uriStr = bundle.getString(KEY_URI, "");
            if (!uriStr.isEmpty()) result = Uri.parse(uriStr);

            return result;
        }

        public static List<Attribute> parseSearchUri(Uri uri) {
            List<Attribute> result = new ArrayList<>();

            if (uri != null)
                for (String typeStr : uri.getQueryParameterNames()) {
                    AttributeType type = AttributeType.searchByName(typeStr);
                    if (type != null)
                        for (String attrStr : uri.getQueryParameters(typeStr)) {
                            String[] attrParams = attrStr.split(";");
                            if (3 == attrParams.length) {
                                result.add(new Attribute(type, attrParams[1]).setId(Long.parseLong(attrParams[0])).setExcluded(Boolean.parseBoolean(attrParams[2])));
                            }
                        }
                }

            return result;
        }
    }


}

File: app/src/main/java/me/devsaki/hentoid/activities/bundles/SearchActivityBundle.kt
package me.devsaki.hentoid.activities.bundles

import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import androidx.core.net.toUri
import me.devsaki.hentoid.database.domains.Attribute
import me.devsaki.hentoid.database.domains.AttributeMap
import me.devsaki.hentoid.enums.AttributeType
import me.devsaki.hentoid.util.Location
import me.devsaki.hentoid.util.SearchCriteria
import me.devsaki.hentoid.util.Type
import me.devsaki.hentoid.util.boolean
import me.devsaki.hentoid.util.intArrayList
import me.devsaki.hentoid.util.long
import me.devsaki.hentoid.util.string

private const val ATTR_EXCLUDED_TYPES = "excludedTypes"
private const val ATTR_LOCATION = "location"
private const val ATTR_CONTENT_TYPE = "contentType"

/**
 * Helper class to transfer data from any Activity to [me.devsaki.hentoid.activities.SearchActivity]
 * through a Bundle
 */
class SearchActivityBundle(val bundle: Bundle = Bundle()) {

    // Only used internally by SearchActivity to communicate with its bottom fragment
    var attributeTypes by bundle.intArrayList()

    // Used both internally and externally (i.e. to communicate with LibraryActivity) for display
    var excludeMode by bundle.boolean(default = false)

    // Used both internally and externally (i.e. to communicate with LibraryActivity) for actual search
    var groupId by bundle.long(default = -1)

    var uri: String by bundle.string(default = "")


    // Helper methods
    companion object {
        fun buildSearchUri(
            searchCriteria: SearchCriteria,
            query: String? = null
        ): Uri {
            return buildSearchUri(
                searchCriteria.attributes,
                searchCriteria.excludedAttributeTypes,
                query ?: searchCriteria.query,
                searchCriteria.location.value,
                searchCriteria.contentType.value
            )
        }

        fun buildSearchUri(
            attributes: Set<Attribute>?,
            excludedTypes: Collection<AttributeType>? = null,
            query: String = "",
            location: Int = 0,
            contentType: Int = 0
        ): Uri {
            val searchUri = Uri.Builder()
                .scheme("search")
                .authority("hentoid")

            if (query.isNotEmpty()) searchUri.path(query)

            if (attributes != null) addAttrs(attributes, searchUri)

            if (excludedTypes != null) addAttrTypeExclusion(excludedTypes, searchUri)

            if (location > 0) searchUri.appendQueryParameter(ATTR_LOCATION, location.toString())
            if (contentType > 0) searchUri.appendQueryParameter(
                ATTR_CONTENT_TYPE,
                contentType.toString()
            )

            return searchUri.build()
        }

        private fun addAttrs(attributes: Set<Attribute>, uri: Uri.Builder) {
            val metadataMap = AttributeMap()
            metadataMap.addAll(attributes)

            for ((attrType, attrs) in metadataMap) {
                for (attr in attrs)
                    uri.appendQueryParameter(
                        attrType.name,
                        "${attr.id};${attr.name};${attr.isExcluded}"
                    )
            }
        }

        private fun addAttrTypeExclusion(excludedTypes: Collection<AttributeType>, uri: Uri.Builder) {
            uri.appendQueryParameter(
                ATTR_EXCLUDED_TYPES,
                TextUtils.join(";", excludedTypes.map { it.code })
            )
        }

        fun parseSearchUri(uri: String): SearchCriteria {
            return parseSearchUri(uri.toUri())
        }

        fun parseSearchUri(uri: Uri): SearchCriteria {
            val attrs: MutableSet<Attribute> = HashSet()
            val excludedTypes: MutableSet<AttributeType> = HashSet()
            var location = 0
            var contentType = 0
            var query = uri.path ?: ""
            // Remove the leading '/'
            if (query.isNotEmpty()) query = query.substring(1)
            for (typeStr in uri.queryParameterNames) {
                val type = AttributeType.searchByName(typeStr)
                if (type != null) { // Parameter is an Attribute
                    for (attrStr in uri.getQueryParameters(typeStr)) {
                        val attrParams = attrStr.split(";").toTypedArray()
                        if (3 == attrParams.size) {
                            val attr = Attribute(
                                type = type,
                                name = attrParams[1],
                                dbId = attrParams[0].toLong()
                            )
                            attr.isExcluded = attrParams[2].toBoolean()
                            attrs.add(attr)
                        }
                    }
                } else {
                    if (ATTR_EXCLUDED_TYPES == typeStr)
                        excludedTypes.addAll(
                            uri.getQueryParameters(typeStr)[0].split(";")
                                .mapNotNull { AttributeType.searchByCode(it.toInt()) }
                        )
                    if (ATTR_LOCATION == typeStr)
                        location = uri.getQueryParameters(typeStr)[0].toInt()
                    if (ATTR_CONTENT_TYPE == typeStr)
                        contentType = uri.getQueryParameters(typeStr)[0].toInt()
                }
            }
            return SearchCriteria(
                query,
                attrs,
                excludedTypes,
                Location.fromValue(location),
                Type.fromValue(contentType)
            )
        }
    }
}